text/0061-functions.md

```diff
@@ -0,0 +1,295 @@
+
+## Detailed Design
+
+### Function declarations
+
+This RFC adds a new top level form to Cedar policysets: the function declaration.
+
+A Cedar function declaration is composed of three elements:
+
+1. A name, which is a valid (possibly namespaced) identifier.
+2. A list of parameters, each of which is a non-namespaced identifier.
+3. A body, which is a Cedar expression.
+    1. The body may contain (non-namespaced) variables, drawn from the function's parameters.
+    2. The body may not contain calls to other functions.
+
+
+Structurally, a declaration is written like this:
+
+```
+function name(param1, param2) {
+  body
+};
+```
+Standard Cedar variables (`principal`, `action`, `resource`, and `context`) are *not* considered bound with the body. 
+(Following the principal of macro hygiene)
+Use of an unbound variable in the body is a syntax error.
+A parameter list may not declare the same variable twice, and may not list any standard Cedar variables.
+An unused variable is a syntax warning.
+Function and variable names share the same namesapce, with standard lexical scoping.
+Inside of a function, any function application (see below) that does not resolve to an extension function or built-in operation is an error. In other words, Cedar functions are not permitted to call other Cedar functions.
+
+### Function applications (a.k.a. function calls)
+
+A function application has the same syntax as an extension function constructor application. In particular, a function application is composed of two elements:
+
+1. The function name (potentially namespaced)
+2. A comma separated list of arguments
+
+Here is an example:
+
+```
+foo(1,2, principal.name)
+```
+
+Function arguments are eagerly evaluated (i.e., call-by-value), as with extension functions today.
+Functions do not exist at run-time -- they cannot be stored in entity attributes, requests, etc. As such, referencing a function by its name, without calling it, is a syntax error.
+Arguments for each of a function's declared parameters, and no more, must be provided at the call, or it's a syntax error.
+Other errors (such as type errors or overflow) are detected at runtime, or via validation.
+Examples:
+```
+function foo(a, b) {
+    a + b
+};
+
+permit(principal,action,resource) when {
+  foo + 1 // Parse error
+};
+permit(principal,action,resource) when {
+    foo(1) // Parse error
+};
+permit(principal,action,resource) when {
+    foo(1, 2) // No error
+};
+permit(principal,action,resource) when {
+    foo(1, "hello") // Run time type error
+};
+permit(principal,action,resource) when {
+    foo(1, "hello", principal) // Parse error
+};
+permit(principal,action,resource) when {
+    bar(1, "hello", principal) // Parse error
+};
```

Comment:
This implies that the parser is stateful in a way it is not currently today.  In order to determine whether these examples are parse errors, the parser has to refer to information it has already parsed earlier in the file (the definition of `foo`).  I'm not sure this is the design we want.

An alternative would be that calling a nonexistent function, or calling with the wrong number of arguments, is not a parse error but an evaluation error, and also detected by validation of course.

Edit: With the below stipulation that function declarations do not have to lexically precede use, this is even more problematic.

Reply:
The argument against making them runtime errors is that that reifies functions. Under the current proposal, functions could be fully erased.

Reply:
Can you not do the checking in two steps? Basically:
1. While processing an input file, store all function defs in a global table that logs name, args, body. For each function application, also consult the table for consistency. If the function is not yet in the table, then add it with the arity and "none" for the body. If a def comes after a call, check it for consistency and replace "none".
2. Make sure no functions in the table have "none" for the body.

Doesn't sound hard?

Reply:
I don't think it's hard, I think Craig's point is that until now we have not had to deal with any kind of use-def relationships in the parser, and this RFC changes that. If we make it a runtime error, our parser stays the way it is.
I think it's worth the change, and I think the RFC doesn't meet our needs if it's a runtime error, but the callout is correct.